std/range/package.d

```diff
@@ -8606,6 +8606,385 @@ if (isForwardRange!Source && hasLength!Source)
+                _maxLen = maxLen;
+                if (!_source.empty)
+                    _isEmpty = false;
+                else
+                    _isEmpty = true;
+            }
+
+            /// Input range primitives. Always present.
+            @property bool empty()
+            {
+                return _isEmpty;
+            }
+
+            /// Ditto
+            @property auto front()
+            {
+                assert(!empty, "Attempting to fetch the front of an empty range");
+
+                if (_endIndexes.empty && !_source.empty)
+                    return _source.save;
+
+                assert(_endIndexes.front != 0, "Attempting to create a chunk of length 0");
+
+                return _source.save.take(_endIndexes.front()-_offset);
+            }
+
+            /// Ditto
+            void popFront()
+            {
+                assert(!empty, "Attempting to popFront an empty range");
+
+                // source range is longer than index
+                if (_endIndexes.empty && !_source.empty)
+                {
+                    _isEmpty = true;
+                    return;
+                }
+
+                _source.popFrontN(_endIndexes.front()-_offset);
+                _offset = _endIndexes.front();
+                _endIndexes.popFront();
+
+                if (_maxLen)
+                {
+                    _maxLen.decrease();
+                    _isEmpty = _maxLen.empty;
+                }
+
+                // length of the source range is smaller than or equal to the index
+                if (_source.empty)
+                    _isEmpty = true;
+            }
+
+            /// Forward range primitives
+            @property typeof(this) save()
+            {
+                return typeof(this)(_source, _endIndexes, _offset);
+            }
+
+            static if (isInfinite!Source && hasLength!Indexes)
+            /// Length is only provided for infinite sources, when hasLength!Indexes is true
+                @property size_t length()
+                {
+                    return _endIndexes.length + 1;
+                }
+
+            static if (hasSlicing!Source && hasSlicing!Indexes && hasLength!Indexes)
+            {
+                /**
+                Indexing and slicing operations are only provided when both Source and Indexes
+                have slicing and hasLength!Indexes is true.
+                */
+                auto opIndex(size_t index)
+                {
+                    assert(index<=_endIndexes.length, "Chunks index out of bounds");
+
+                    const size_t startIndex = (index == 0 ? 0 : _endIndexes[index-1] - _offset);
+
+                    static if (hasLength!Source)
+                        assert(startIndex<=_source.length, "Chunks index out of bounds");
+
+                    if (index == _endIndexes.length)
+                        return _source[startIndex .. $];
+
+                    static if (hasLength!Source)
```

Comment:
Took me a bit to realize that this can only be false if Source is infinite. Maybe add a comment.